js/src/evaluation/llm.ts

```diff
@@ -0,0 +1,295 @@
+ * import { OpenAI } from "@langchain/openai"
+ *
+ *
+ * const evaluator = new LLMEvaluator({
+ *   promptTemplate: "Rate the quality of this response...",
+ *   scoreConfig: new ContinuousScoreConfig({
+ *     key: "quality",
+ *     description: "Quality score from 0-1",
+ *     min: 0,
+ *     max: 1
+ *   }),
+ *   chatModel: new OpenAI({ model: "gpt-4" })
+ * });
+ * ```
+ *
+ * @implements {RunEvaluator}
+ */
+export class LLMEvaluator implements RunEvaluator {
+  prompt: any;
+  mapVariables?: (run: Run, example?: Example) => Record<string, any>;
+  scoreConfig: ScoreConfig;
+  scoreSchema: Record<string, any>;
+  runnable: any;
+
+  constructor(params: LLMEvaluatorParams) {
+    const { promptTemplate, scoreConfig, chatModel, mapVariables } = params;
+
+    // Store the configuration
+    this.scoreConfig = scoreConfig;
+    this.mapVariables = mapVariables;
+
+    // Create the score schema
+    this.scoreSchema = createScoreJsonSchema(scoreConfig);
+
+    // Create the prompt template
+    if (typeof promptTemplate === "string") {
+      this.prompt = ChatPromptTemplate.fromMessages([
+        { role: "human", content: promptTemplate },
+      ]);
+    } else {
+      this.prompt = ChatPromptTemplate.fromMessages(promptTemplate);
+    }
+
+    const modelWithStructuredOutput = chatModel.withStructuredOutput
+      ? chatModel.withStructuredOutput(this.scoreSchema)
+      : null;
+    if (!modelWithStructuredOutput) {
+      throw new Error("Passed chat model must support structured output");
+    }
+    this.runnable = this.prompt.pipe(modelWithStructuredOutput);
+  }
+
+  async evaluateRun(
+    run: Run,
+    example?: Example
+  ): Promise<EvaluationResult | EvaluationResults> {
+    const runId = uuid.v4();
+    const variables = this.prepareVariables(run, example);
+    const output = await this.runnable.invoke(variables, { runId: runId });
+
+    return this.parseOutput(output, runId);
+  }
+
+  private prepareVariables(run: Run, example?: Example): Record<string, any> {
+    if (this.mapVariables) {
+      return this.mapVariables(run, example);
+    }
+
+    const variables: Record<string, any> = {};
+
+    // Input handling
+    if (Object.keys(run.inputs).length === 0) {
+      throw new Error(
+        "No input keys are present in run.inputs but the prompt requires 'input'."
+      );
+    }
+    if (Object.keys(run.inputs).length !== 1) {
+      throw new Error(
+        "Multiple input keys are present in run.inputs. Please provide a map_variables function."
+      );
+    }
+    variables.input = Object.values(run.inputs)[0];
+
+    // Output handling
+    if (!run.outputs || Object.keys(run.outputs).length === 0) {
+      throw new Error(
+        "No output keys are present in run.outputs but the prompt requires 'output'."
+      );
+    }
+    if (Object.keys(run.outputs).length !== 1) {
+      throw new Error(
+        "Multiple output keys are present in run.outputs. Please provide a map_variables function."
```

Comment:
This uses the same function as inputs? Is that a typo?

Sidebar: I generally dislike this `keys` pattern but maybe it's more popular in Python